### PR TITLE
Refactor: implement History as an interface

### PR DESCRIPTION
**Pull request by Funfun**
Hey @d4l3k

In my spare time, I checked out your project. Really thanks for doing it! Here is a bit of my refactoring. The motivation was to make the code a bit structured and I picked a history functionality due to the simplicity of it. Hope you would find it useful. 

NOTE: i did not rename `io_default.go`, `io_js.go` to keep the diff understable  and clear.

### CODE REVIEW

File: pry/io_default.go [hunk 1]
```diff
@@ -15,43 +15,67 @@ var readFile = ioutil.ReadFile
 
 var historyFile = ".go-pry_history"
 
-func historyPath() (string, error) {
-	dir, err := homedir.Dir()
-	if err != nil {
-		return "", err
-	}
-	return path.Join(dir, historyFile), nil
+type History interface {
```
[the remark this answers is not in the document]
Comment (d4l3k) [reply pry/io_default.go]: Yeah, sounds good! Shouldn't be too hard to add it down the road.

File: pry/io_default.go [hunk 2]
```diff
@@ -15,43 +15,67 @@ var readFile = ioutil.ReadFile
...
+}
+
+func NewHistory() *IOHistory {
+	h := IOHistory{}
+	h.FileName = historyFile
+
+	dir, err := homedir.Dir()
 	if err != nil {
 		log.Printf("Error finding user home dir: %s", err)
 		return nil
```
[the remark this answers is not in the document]
Comment (Funfun) [reply pry/io_default.go]: yeah, that actually would be more elegant. On top of that, after a call, down on call stack, we can log with an error message and remove logging from here.

**Comment by Funfun** [comment]
need to test js version

File: pry/io_default.go [hunk 3]
```diff
@@ -15,43 +15,63 @@ var readFile = ioutil.ReadFile
...
 }
 
-func saveHistory(history *[]string) {
-	body, err := json.Marshal(history)
+// Save saves marshaled history's records into file
+func (h IOHistory) Save() error {
+	body, err := json.Marshal(h.Records)
 	if err != nil {
 		log.Printf("Err marshalling history: %s", err)
+		return err
```
Comment (d4l3k) [inline pry/io_default.go]: could you change this to just be 
```
return errors.Wrapf(err, "error marshaling history")
```
and get rid of the print statement?

File: pry/io_default.go [hunk 4]
```diff
@@ -15,43 +15,63 @@ var readFile = ioutil.ReadFile
...
 	}
-	path, err := historyPath()
-	if err != nil {
-		log.Printf("Error finding user home dir: %s", err)
-		return
-	}
-	if err := ioutil.WriteFile(path, body, 0755); err != nil {
+	if err := ioutil.WriteFile(h.FilePath, body, 0755); err != nil {
 		log.Printf("Error writing history: %s", err)
+		return err
```
Comment (d4l3k) [inline pry/io_default.go]: same thing here

File: pry/io_js.go [hunk 5]
```diff
@@ -31,22 +31,50 @@ func readFile(path string) ([]byte, error) {
...
 	hist := js.Global().Get("localStorage").Get("history")
 	if hist.Type() == js.TypeUndefined {
+		// dirty because it is actually an error
 		return nil
 	}
-	var history []string
-	if err := json.Unmarshal([]byte(hist.String()), &history); err != nil {
+	var records []string
+	if err := json.Unmarshal([]byte(hist.String()), &records); err != nil {
 		panic(err)
```
Comment (d4l3k) [inline pry/io_js.go]: return an error instead of panicking

File: pry/io_js.go [hunk 6]
```diff
@@ -31,22 +31,50 @@ func readFile(path string) ([]byte, error) {
...
 	}
-	return history
+	bh.Records = records
+
+	return true
```
Comment (d4l3k) [inline pry/io_js.go]: think this should be nil

File: pry/io_js.go [hunk 7]
```diff
@@ -31,22 +31,50 @@ func readFile(path string) ([]byte, error) {
...
 }
 
-func saveHistory(history *[]string) {
-	bytes, err := json.Marshal(history)
+// Save saves marshaled history's records into localStorage
+func (bh BrowserHistory) Save() error {
+	bytes, err := json.Marshal(bh.Records)
 	if err != nil {
 		panic(err)
```
Comment (d4l3k) [inline pry/io_js.go]: return error

File: pry/pry.go [hunk 8]
```diff
@@ -52,12 +52,18 @@ func apply(
 
 	displayFilePosition(out, filePathRaw, filePath, lineNum)
 
-	history := loadHistory()
+	history, err := NewHistory()
+	if err != nil {
+		fmt.Errorf("Failed to initiliaze history %+v", err)
+	}
+	if err := history.Load(); err != nil {
+		fmt.Errorf("Failed to load the history %+v", err)
```
Comment (d4l3k) [inline pry/pry.go]: you can also just return an error here with errors.Wrapf

File: pry/io_js.go (hunk 7 above)
Comment (Funfun) [reply pry/io_js.go]: why you had a `panic` originally?

**Comment by d4l3k** [comment]
Thanks for contributing! Sorry for the delay in merging.

**Pull request merged**